src/recon_buildblock/LogcoshPrior.cxx

```diff
@@ -0,0 +1,666 @@
+        compute_weights(weights, current_image_cast.get_grid_spacing(), this->only_2D);
+      }
+
+      const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+      if (do_kappa && !kappa_ptr->has_same_characteristics(current_image_estimate))
+        error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+      const int min_z = current_image_estimate.get_min_index();
+      const int max_z = current_image_estimate.get_max_index();
+      for (int z=min_z; z<=max_z; z++)
+      {
+        const int min_dz = max(weights.get_min_index(), min_z-z);
+        const int max_dz = min(weights.get_max_index(), max_z-z);
+
+        const int min_y = current_image_estimate[z].get_min_index();
+        const int max_y = current_image_estimate[z].get_max_index();
+
+        for (int y=min_y;y<= max_y;y++)
+        {
+          const int min_dy = max(weights[0].get_min_index(), min_y-y);
+          const int max_dy = min(weights[0].get_max_index(), max_y-y);
+
+          const int min_x = current_image_estimate[z][y].get_min_index();
+          const int max_x = current_image_estimate[z][y].get_max_index();
+          for (int x=min_x;x<= max_x;x++)
+          {
+            const int min_dx = max(weights[0][0].get_min_index(), min_x-x);
+            const int max_dx = min(weights[0][0].get_max_index(), max_x-x);
+
+            elemT gradient = 0;
+            for (int dz=min_dz;dz<=max_dz;++dz)
+              for (int dy=min_dy;dy<=max_dy;++dy)
+                for (int dx=min_dx;dx<=max_dx;++dx)
+                {
+                  // psi'(t)/t = tanh/t
+                  elemT current;
+                  elemT temp = fabs(current_image_estimate[z][y][x] - current_image_estimate[z+dz][y+dy][x+dx]);
+
+                  //if (temp <= 0.01)
+                  //    current = weights[dz][dy][dx];
+                  //else
+                  current = weights[dz][dy][dx] * (1/(this->scalar))*tanh(this->scalar*(temp+0.01))/(temp+0.01);
+
+                  if (do_kappa)
+                    current *=
+                            (*kappa_ptr)[z][y][x] * (*kappa_ptr)[z+dz][y+dy][x+dx];
+
+                  gradient += current;
+                }
+
+            parabolic_surrogate_curvature[z][y][x]= gradient * this->penalisation_factor;
+          }
+        }
+      }
+
+      info(boost::format("parabolic_surrogate_curvature max %1%, min %2%\n") % parabolic_surrogate_curvature.find_max() % parabolic_surrogate_curvature.find_min());
+      /*{
+       static int count = 0;
+       ++count;
+       char filename[20];
+       sprintf(filename, "normalised_gradient%d.v",count);
+       write_basic_interfile(filename, parabolic_surrogate_curvature);
+       }*/
+    }
+
+    template <typename elemT>
+    Succeeded
+    LogcoshPrior<elemT>::
+    add_multiplication_with_approximate_Hessian(DiscretisedDensity<3,elemT>& output,
+                                                const DiscretisedDensity<3,elemT>& input) const
```

Comment:
Possibly not enough arguments. In theory this would requires 3 arguments (`output, input, current_image_estimate`). In our planned usage of this method, `input = current_image_estimate`, but what if it doesn't? Does this ever exist? QP does not suffer from this problem because x=1 and therefore `current_image_estimate` is ignored and there are only two arguments (`output, input`).